### PR TITLE
Add recipe for OpenMM

### PR DESCRIPTION
**Pull request by jaimergp**
Trying to build OpenMM in conda-forge with GPU support, following examples at #8637, #8638.

__Checklist__

- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [X] Source is from official source
- [x] Package does not vend other packages
- [x] Build number is 0
- [X] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your
      recipe (see
      [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos)
      for more details)
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

### CODE REVIEW

**Comment by jakirkham** [comment]
Thanks for working on this @jaimergp! I'm trying to sort out how we generally handle variants for GPU package builds in issue ( https://github.com/conda-forge/conda-forge-pinning-feedstock/issues/237 ), which should help us get this and other GPU packages in.

**Comment by jaimergp** [comment]
@jakirkham 

Thanks for the tip! Following the issue now!

File: recipes/openmm/build.sh [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    # CUDA_HOME is defined by nvcc metapackage
+    CMAKE_FLAGS+=" -DCUDA_TOOLKIT_ROOT_DIR=${CUDA_HOME}"
+    # From: https://github.com/floydhub/dl-docker/issues/59
+    CMAKE_FLAGS+=" -DCMAKE_LIBRARY_PATH=${CUDA_HOME}/lib64/stubs"
+    # CUDA tests won't build, disable for now
+    # See https://github.com/openmm/openmm/issues/2258#issuecomment-462223634
+    CMAKE_FLAGS+=" -DOPENMM_BUILD_CUDA_TESTS=OFF"
+    # CUDA OpenCL
+    CMAKE_FLAGS+=" -DOPENCL_INCLUDE_DIR=${CUDA_HOME}/include/"
+    CMAKE_FLAGS+=" -DOPENCL_LIBRARY=${CUDA_HOME}/lib64/libOpenCL.so"
```
Comment (isuruf) [inline recipes/openmm/build.sh]: Does this need to be CUDA's OpenCL? You should use an ICD loader. See PyOpenCL recipe for the ICD loader library names.

**Comment by jaimergp** [comment]
Hi @conda-forge/core !

This should be ready for review. We would like to have it merged without producing packages yet though (that's why all builds are disabled), so we can have a feedstock where we can push to a `testing` label or something similar until we are ready to create a stable package. Would that be ok?

Thanks a lot!

**Comment by isuruf** [comment]
You can change the target in staged-recipes as well. No need to wait for a feedstock

**Comment by isuruf** [comment]
Can you remove unnecessary changes like adding Nvidia channel? And also remove disabling error_overlinking

File: conda-forge.yml [hunk 2]
```diff
@@ -1,2 +1,5 @@
 max_py_ver: '37'
-max_r_ver: '35'
\ No newline at end of file
+max_r_ver: '35'
+channels:
+  targets:
+    - ["conda-forge", "testing"]
```
Comment (isuruf) [inline conda-forge.yml]: You shouldn't change this file. You need to add a conda_build_config file to the recipe. See conda-smithy documentation

Comment (jaimergp) [reply conda-forge.yml]: Oops, sorry about that. Thanks!

File: .circleci/run_docker_build.sh [hunk 3]
```diff
@@ -3,7 +3,7 @@
 # NOTE: This script has been adapted from content generated by github.com/conda-forge/conda-smithy
 
 REPO_ROOT=$(cd "$(dirname "$0")/.."; pwd;)
-IMAGE_NAME="condaforge/linux-anvil-comp7"
+IMAGE_NAME="condaforge/linux-anvil-cuda:10.0"
```
Comment (jakirkham) [inline .circleci/run_docker_build.sh]: So this is fine for testing purposes, but we will need to revert this before merging.

File: .circleci/conda_build_config.yaml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+nvcc_compiler:
+  - nvcc
+nvcc_compiler_version:
+  - 10.0
```
Comment (jakirkham) [inline .circleci/conda_build_config.yaml]: Same story here.

Comment (jaimergp) [reply .circleci/conda_build_config.yaml]: This is handled by `conda-smithy` and the right `conda_build_config.yaml` fields now, right?

File: recipes/openmm/conda_build_config.yaml [hunk 5]
```diff
@@ -0,0 +1,6 @@
+cuda_compiler:
+- nvcc
+cuda_compiler_version:
+- 10.0
```
Comment (jakirkham) [inline recipes/openmm/conda_build_config.yaml]: Do you support other CUDA versions? Or only 10.0?

Also do you want to have a CPU only build? Or are you only interested in building with GPU support?

File: recipes/openmm/conda_build_config.yaml [hunk 6]
```diff
@@ -0,0 +1,6 @@
...
+channel_targets:
+- conda-forge testing
```
Comment (jakirkham) [inline recipes/openmm/conda_build_config.yaml]: What is this for?

File: recipes/openmm/conda_build_config.yaml (hunk 5 above)
Comment (jaimergp) [reply recipes/openmm/conda_build_config.yaml]: We will add more CUDA versions, yes. I think we can do it [like the ucx recipe](https://github.com/conda-forge/ucx-split-feedstock/blob/master/recipe/conda_build_config.yaml), right?

OpenMM will just ignore CUDA/OpenCL if the platform is not compatible, so we don't _need_ CPU-only packages, I'd say. We might be interested in building variants without the explicit `cudatoolkit` runtime dependency though, so people can use their own CUDA installation without downloading a several hundreds MB package. How are you handling this scenario right now?

File: recipes/openmm/conda_build_config.yaml (hunk 6 above)
Comment (jaimergp) [reply recipes/openmm/conda_build_config.yaml]: We don't want packages in `conda-forge/main` without manually testing them with an actual GPU. Given the current OpenMM [docs](http://docs.openmm.org/7.4.0/userguide/application.html#installing-openmm), `conda-forge` has precedence over `omnia` (where `openmm` resides now), and any new users that follow those instructions would obtain an untested (potentially broken) package.

**Comment by jakirkham** [comment]
Thanks for continuing to push this forward, @jaimergp. Added a couple comments to guide the next steps here.

File: .circleci/conda_build_config.yaml (hunk 4 above)
Comment (jaimergp) [reply .circleci/conda_build_config.yaml]: Fixed!

File: .circleci/run_docker_build.sh (hunk 3 above)
Comment (jaimergp) [reply .circleci/run_docker_build.sh]: Fixed!

Comment (isuruf) [reply .circleci/run_docker_build.sh]: You need to keep this, remove skipping, make tests pass. Once that is reviewed, we can revert this change and merge.

File: recipes/openmm/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+  url: https://github.com/openmm/{{ name }}/archive/{{ version }}.tar.gz
+  sha256: cb7ddd923d6f402cf2723b3c14509915abad328505e8ce401ac3131706dabeff
+
+build:
+  number: {{ build }}
+  # skip: True  # [py2k or win]
+  skip: True
+  missing_dso_whitelist:
+  - "*/libcuda.*"  # [linux]
+  - "*/libOpenCL.*"  # [unix]
```
Comment (isuruf) [inline recipes/openmm/meta.yaml]: Remove this line.

File: recipes/openmm/conda_build_config.yaml (hunk 6 above)
Comment (jakirkham) [reply recipes/openmm/conda_build_config.yaml]: Why does `conda-forge` have precedence over `omnia`? Would figure it would be the other way around?

File: recipes/openmm/conda_build_config.yaml (hunk 5 above)
Comment (jakirkham) [reply recipes/openmm/conda_build_config.yaml]: By adding a [`None` version]( https://github.com/conda-forge/ucx-split-feedstock/blob/e807987cd0a8d36ffdd2d4edb41c80442cef8b17/recipe/conda_build_config.yaml#L4 ) and [a selector to disable the CUDA compiler in that case]( https://github.com/conda-forge/ucx-split-feedstock/blob/e807987cd0a8d36ffdd2d4edb41c80442cef8b17/recipe/meta.yaml#L50 ).

File: .circleci/conda_build_config.yaml (hunk 4 above)
Comment (jakirkham) [reply .circleci/conda_build_config.yaml]: > This is handled by `conda-smithy` and the right `conda_build_config.yaml` fields now, right?

Yeah that's right.

File: recipes/openmm/conda_build_config.yaml (hunk 6 above)
Comment (Lnaden) [reply recipes/openmm/conda_build_config.yaml]: A while back (a couple years ago now) when we were planning all this out, we wanted to eventually move every package over from `omnia` to `conda-forge`, so us `omnia` devs could eventually stop having to maintain a separate conda-based build system ourselves. The hard part of that was going to be resolving all the dependencies, which we had to maintain/copy a large number of them over to the omnia channel. 

To prepare for this migration, we decided that `conda-forge` would be the de-facto dependency source, and we only maintained on `omnia` the minimum set of packages we had to. This basically meant that the packages which were were all OpenMM dependent at some level. The goal was, once OpenMM came to `conda-forge`, it would be very easy for all the other packages to migrate over as well since all they would have to do (in theory) was create a `staged-recipe` and the dependencies would basically already be resolved since we were relying on `conda-forge` as the higher priority channel anyways.

File: recipes/openmm/conda_build_config.yaml (hunk 5 above)
Comment (jaimergp) [reply recipes/openmm/conda_build_config.yaml]: But we still want to build the CUDA libraries in our package. We just want to ignore `cudatoolkit` for one of the variants so the user provides their own CUDA install (like in a cluster).

File: .circleci/run_docker_build.sh (hunk 3 above)
Comment (jaimergp) [reply .circleci/run_docker_build.sh]: Ok!

File: recipes/openmm/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+source:
+  url: https://github.com/openmm/{{ name }}/archive/{{ version }}.tar.gz
+  sha256: cb7ddd923d6f402cf2723b3c14509915abad328505e8ce401ac3131706dabeff
+
+build:
+  number: {{ build }}
+  skip: True  # [py2k or win]
+  missing_dso_whitelist:
+  - "*/libcuda.*"  # [linux]
```
Comment (jakirkham) [inline recipes/openmm/meta.yaml]: Why is this needed?

File: recipes/openmm/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+  host:
+    - python
+    - fftw
+    - swig
+    - pthread-stubs
+    - lxml
+    - numpy
+    - cython
+    # needed for Python wrappers
+    - doxygen 1.8.14
```
Comment (jakirkham) [inline recipes/openmm/meta.yaml]: Why is this version needed?

File: recipes/openmm/meta.yaml (hunk 8 above)
Comment (isuruf) [reply recipes/openmm/meta.yaml]: Because we've enabled `error_overlinking` on staged-recipes and since `/usr/local/lib/libcuda.so` is outside of prefix, conda-build errors. Correct way to fix this would be to generate the stub library `libcuda.so` ourselves and make that a CDT.

File: recipes/openmm/meta.yaml [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+    # OpenCL ICD
+    - ocl-icd  # [linux]
+    - khronos-opencl-icd-loader  # [osx or win]
+
+  run:
+    - python
+    - fftw
+    - numpy
+    - cython
```
Comment (jakirkham) [inline recipes/openmm/meta.yaml]: Why is `cython` needed at run time?

File: recipes/openmm/conda_build_config.yaml (hunk 5 above)
Comment (jakirkham) [reply recipes/openmm/conda_build_config.yaml]: Ah. I'm not sure that we have a good way to solve this ATM. What is gained by using an existing CUDA install vs. using the `cudatoolkit` package?

File: recipes/openmm/meta.yaml [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+    (especially on recent GPUs) that make it truly unique among simulation
+    codes.
+  doc_url: http://docs.openmm.org
+  dev_url: https://github.com/openmm/openmm
+
+extra:
+  recipe-maintainers:
+    - jchodera
+    - jaimergp
+    - peastman
```
Comment (jakirkham) [inline recipes/openmm/meta.yaml]: @jchodera @peastman , are you ok being maintainers on `openmm`?

File: .circleci/run_docker_build.sh (hunk 3 above)
Comment (isuruf) [reply .circleci/run_docker_build.sh]: Since you are using 10.0 image, you should skip other versions in staged-recipes and only enable them in the feedstock.

**Comment by jakirkham** [comment]
Had a couple questions about the recipe generally, but I think this is largely looking good at this point. Thanks for your efforts and patience @jaimergp as we figure this out together. 🙂

File: recipes/openmm/meta.yaml (hunk 11 above)
Comment (jchodera) [reply recipes/openmm/meta.yaml]: Yes!

File: recipes/openmm/meta.yaml (hunk 8 above)
Comment (jakirkham) [reply recipes/openmm/meta.yaml]: Can you please raise an issue about this somewhere so we can follow-up? Maybe the `nvcc` feedstock or the webpage repo?

File: recipes/openmm/meta.yaml (hunk 11 above)
Comment (peastman) [reply recipes/openmm/meta.yaml]: Yes.

File: recipes/openmm/meta.yaml (hunk 10 above)
Comment (peastman) [reply recipes/openmm/meta.yaml]: No reason I know of!

Comment (jakirkham) [reply recipes/openmm/meta.yaml]: Let's go ahead and drop it then.

```suggestion
```

File: recipes/openmm/meta.yaml (hunk 8 above)
Comment (jakirkham) [reply recipes/openmm/meta.yaml]: This is now being tracked in issue ( https://github.com/conda-forge/nvcc-feedstock/issues/12 ).

File: recipes/openmm/meta.yaml (hunk 9 above)
Comment (jchodera) [reply recipes/openmm/meta.yaml]: @peastman: Do you recall which Doxygen versions work correctly with the OpenMM Swig system?

Comment (peastman) [reply recipes/openmm/meta.yaml]: I don't know if this was in response to a particular problem.  In general, every time we change versions of Doxygen or SWIG there's a good chance something will break.  So when possible it's best to pin the build tools to particular versions we know work.  We also should be pinning SWIG, since we know there's a bug in recent versions that breaks the OpenMM documentation (https://github.com/sphinx-doc/sphinx/issues/6565).

File: recipes/openmm/meta.yaml (hunk 10 above)
Comment (jaimergp) [reply recipes/openmm/meta.yaml]: Do you agree @jchodera? This was copied from the `omnia` recipe, so I don't know if there's a specific reason to keep it or was simply blindly inherited from previous versions.

File: recipes/openmm/meta.yaml (hunk 9 above)
Comment (jaimergp) [reply recipes/openmm/meta.yaml]: Latest `doxygen` is `v1.8.16`, which breaks the build. `1.8.14` works correctly, that's why it's pinned. Maybe we can relax the pinning a bit (`<=`?). Regarding SWIG, we are not building the docs in this recipe, so we are not observing that bug, I'd say.

File: recipes/openmm/conda_build_config.yaml (hunk 5 above)
Comment (jaimergp) [reply recipes/openmm/conda_build_config.yaml]: I'd say there are two main reasons:

1. Avoiding the large `cudatoolkit` download.
2. Allowing HPC sysadmins to provide their own CUDA environment.

Maybe @peastman wants to add something else.

File: recipes/openmm/meta.yaml (hunk 10 above)
Comment (jchodera) [reply recipes/openmm/meta.yaml]: That was probably an accident.

File: recipes/openmm/meta.yaml (hunk 9 above)
Comment (jchodera) [reply recipes/openmm/meta.yaml]: It would be better to be as permissive as we can, @peastman, since we could cause package conflicts if we pin to a single version.

Comment (peastman) [reply recipes/openmm/meta.yaml]: For runtime dependencies we should be permissive.  Build time dependencies don't cause problems for users, and we know from past experience that changes to the build tools have a very high probability of breaking things.

File: recipes/openmm/conda_build_config.yaml (hunk 5 above)
Comment (peastman) [reply recipes/openmm/conda_build_config.yaml]: That sounds right.

Comment (jakirkham) [reply recipes/openmm/conda_build_config.yaml]: Please raise an issue on the [`nvcc` issue tracker]( https://github.com/conda-forge/nvcc-feedstock/issues ). We can discuss this further there.

Comment (jaimergp) [reply recipes/openmm/conda_build_config.yaml]: Done! https://github.com/conda-forge/nvcc-feedstock/issues/14

**Comment by jaimergp** [comment]
I think I applied all suggested changes! Thanks a lot to @jakirkham and @isuruf for their valuable feedback!

Tests are passing and once we get the feedstock we will be able to test the packages locally.

**Comment by isuruf** [comment]
Looks good to me. Please remove https://github.com/conda-forge/staged-recipes/pull/9102/files#diff-ee85f7cbba6676fce7d05c46b9baff16R6 and we can merge this.

**Comment by jaimergp** [comment]
Thank you very much @isuruf! Suggestion applied!

File: recipes/openmm/conda_build_config.yaml [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+cuda_compiler:
+- nvcc
+cuda_compiler_version:
+# - 9.2
+- 10.0
+# - 10.1
+docker_image:
+# - condaforge/linux-anvil-cuda:9.2
+- condaforge/linux-anvil-cuda:10.0
+# - condaforge/linux-anvil-cuda:10.1
```
Comment (isuruf) [inline recipes/openmm/conda_build_config.yaml]: You can uncomment this in the feedstock. If you want to add more CUDA packages through staged-recipes, we'd appreciate some help in getting staged-recipes to play nicely with CUDA packages.

Comment (jakirkham) [reply recipes/openmm/conda_build_config.yaml]: Have been meaning to raise an issue about this. Submitted issue ( https://github.com/conda-forge/staged-recipes/issues/9773 ) for discussion and tracking work on this task.

Comment (jaimergp) [reply recipes/openmm/conda_build_config.yaml]: Following issue!
